### PR TITLE
LGA-2386 FALA end-to-end test for search by organisation name

### PR DESCRIPTION
**Pull request by MishNajam**
## What does this pull request do?
This PR adds an end-to-end test for FALA. 

It covers the "Search by organisation" scenario:

**Given** I am on the Find a legal aid adviser homepage
**And** I provide location details 
**And** I provide an organisation name 
**When** I select the Search button
**Then** I am taken to the search results page
**And** one result is visible on the results page

### CODE REVIEW

File: behave/features/steps/fala_end_to_end_tests.py [hunk 1]
```diff
@@ -42,3 +50,34 @@ def step_impl_result_page(context, location):
...
+        '//div[@class="search-results-container"]'
+    )
+    result_number_paragraph = context.helperfunc.find_by_xpath(
+        '//p[@class="govuk-body"]'
+    )
+
+    assert (
+        current_path
+    ), f"""{CLA_FALA_URL}/?postcode={location}+&name={organisation}&search="""
+    assert title_xpath, f"{FALA_HEADER}"
```
Comment (said-moj) [inline behave/features/steps/fala_end_to_end_tests.py:71]: What is this assert here? 
Is it assert title_xpath and f"{FALA_HEADER}" are both truthy?

Comment (MishNajam) [reply behave/features/steps/fala_end_to_end_tests.py:71]: my assumption was that this asserts they are both the same/equal. Is this not the correct syntax for this? 🤔

Comment (sarahyjja) [reply behave/features/steps/fala_end_to_end_tests.py:71]: 💡 Maybe this one can be rewritten as `title_xpath` is equal to `{FALA_HEADER} + location + organisation` and not only comparing it to the first part of the `url` ?

Comment (said-moj) [reply behave/features/steps/fala_end_to_end_tests.py:71]: I think it's `assert condition, [message to show when assert fails]`

File: behave/features/steps/fala_end_to_end_tests.py [hunk 2]
```diff
@@ -10,15 +10,25 @@ def step_impl_homepage(context):
...
-    assert input_xpath.get_attribute("value"), location
+    input_value = input_xpath.get_attribute("value")
+    assert input_value == location
+
+
+@step('I provide an organisation name "{organisation}"')
+def step_impl_input_organisation(context, organisation):
+    input_id = context.helperfunc.find_by_id("id_name")
+    input_id.send_keys(organisation)
+    input_xpath = context.helperfunc.find_by_xpath('//input[@id="id_name"]')
```
Comment (psweeting1) [inline behave/features/steps/fala_end_to_end_tests.py:29]: You've already have a variable called `input_id` which has found the id "id_name" so you probably don't need this xpath anymore.

File: behave/features/steps/fala_end_to_end_tests.py [hunk 3]
```diff
@@ -88,4 +102,42 @@ def step_impl_update_result_page(context, location, filter_label):
...
+
+    assert (
+        current_url
+        == f"""{CLA_FALA_URL}/?postcode={location}&name={organisation}&search="""
+    )
+    assert title_xpath == f"{FALA_HEADER}"
+    assert result_container_xpath, result_number_paragraph is not None
+
+
+@step('"{count}" result is visible on the results page')
```
Comment (psweeting1) [inline behave/features/steps/fala_end_to_end_tests.py:135]: We can use Pre-defined Data types https://www.tutorialspoint.com/behave/behave_data_types.htm

So it would be `{count:d}` That accepts a digital which in this case is a number. No need to convert a String to an integer. It's something we don't take advantage of much.

File: behave/features/steps/fala_end_to_end_tests.py [hunk 4]
```diff
@@ -88,4 +102,42 @@ def step_impl_update_result_page(context, location, filter_label):
...
+def step_impl_count_results_visible_on_results_page(context, count):
+    count = int(count)
+    listitems = context.helperfunc.find_many_by_xpath(
+        "//main/div/div[3]/div[1]/section/div/div[2]/div/ul/li"
```
Comment (psweeting1) [inline behave/features/steps/fala_end_to_end_tests.py:139]: This targets a single list item in an under ordered list `<ul>` and not a list of list items `<li>`.

You can do find_many_by_xpath('//ul[@class="org-list"]/li`) which gets a list. Then perform a for loop through the list to count how many list items you have on a page. 

https://stackoverflow.com/questions/52092180/select-li-elements-from-ul-with-xpath Example of someone using a for loop on list items.

File: behave/features/steps/fala_end_to_end_tests.py [hunk 5]
```diff
@@ -68,18 +77,23 @@ def step_impl_apply_filter(context):
...
+    title_xpath = context.helperfunc.find_by_xpath(
+        "//html/body/div/main/div/div/h1"
+    ).text.replace("\n", " ")
     result_container_xpath = context.helperfunc.find_by_xpath(
         '//div[@class="search-results-container"]'
     )
     updated_result_number_paragraph = context.helperfunc.find_by_xpath(
         '//p[@class="govuk-body"]'
     )
+    location = location.replace(" ", "+")
```
Comment (psweeting1) [inline behave/features/steps/fala_end_to_end_tests.py:90]: Naming a variable the same as a parameter is confusing. Can rename the variable please.

File: behave/features/steps/fala_end_to_end_tests.py (hunk 3 above)
Comment (MishNajam) [reply behave/features/steps/fala_end_to_end_tests.py:135]: Added

File: behave/features/steps/fala_end_to_end_tests.py (hunk 4 above)
Comment (MishNajam) [reply behave/features/steps/fala_end_to_end_tests.py:139]: I've tested this solution and it definitely returns a (python) list of <li> elements, which is why I just use the len() function to count the number of <li> elements. (Happy to demo) I've replace the line with this though '//ul[@Class="org-list"]/li` as it is a far neater and precise solution.

File: behave/features/steps/fala_end_to_end_tests.py (hunk 5 above)
Comment (MishNajam) [reply behave/features/steps/fala_end_to_end_tests.py:90]: renamed here and in the other similar steps

File: behave/features/steps/fala_end_to_end_tests.py (hunk 2 above)
Comment (MishNajam) [reply behave/features/steps/fala_end_to_end_tests.py:29]: good spot, amended this pattern in this step + location step

File: behave/features/steps/fala_end_to_end_tests.py [hunk 6]
```diff
@@ -88,4 +103,40 @@ def step_impl_update_result_page(context, location, filter_label):
...
+    )
+    organisation_url_string = organisation.replace(" ", "+")
+    location_url_string = location.replace(" ", "+")
+
+    assert (
+        current_url
+        == f"""{CLA_FALA_URL}/?postcode={location_url_string}&name={organisation_url_string}&search="""
+    )
+    assert title_xpath == f"{FALA_HEADER}"
+    assert result_container_xpath, result_number_paragraph is not None
```
Comment (said-moj) [inline behave/features/steps/fala_end_to_end_tests.py:133]: Should this be two asserts?

File: behave/features/steps/fala_end_to_end_tests.py [hunk 7]
```diff
@@ -88,4 +103,40 @@ def step_impl_update_result_page(context, location, filter_label):
...
+    title_xpath = context.helperfunc.find_by_xpath(
+        "//html/body/div/main/div/div/h1"
+    ).text.replace("\n", " ")
+    result_container_xpath = context.helperfunc.find_by_xpath(
+        '//div[@class="search-results-container"]'
+    )
+    organisation_url_string = organisation.replace(" ", "+")
+    location_url_string = location.replace(" ", "+")
+
+    assert (
```
Comment (said-moj) [inline behave/features/steps/fala_end_to_end_tests.py:128]: Is the parenthesis intentional for this assert?

File: behave/features/steps/fala_end_to_end_tests.py [hunk 8]
```diff
@@ -88,4 +103,40 @@ def step_impl_update_result_page(context, location, filter_label):
 def step_impl_error_shown_on_page(context):
     alert = context.helperfunc.find_by_css_selector(".alert-message")
     assert alert is not None
-    assert alert.text, "No results"
+    assert (
+        alert.text == "No results\nThere are no results matching your search criteria."
```
Comment (said-moj) [inline behave/features/steps/fala_end_to_end_tests.py:107]: Is the parentheses intentional for this assert?

File: behave/features/steps/fala_end_to_end_tests.py [hunk 9]
```diff
@@ -68,18 +78,23 @@ def step_impl_apply_filter(context):
...
+    ).text.replace("\n", " ")
     result_container_xpath = context.helperfunc.find_by_xpath(
         '//div[@class="search-results-container"]'
     )
     updated_result_number_paragraph = context.helperfunc.find_by_xpath(
         '//p[@class="govuk-body"]'
     )
+    location_url_string = location.replace(" ", "+")
+
     assert (
```
Comment (said-moj) [inline behave/features/steps/fala_end_to_end_tests.py:93]: Is the parentheses intentional for this assert?

File: behave/features/steps/fala_end_to_end_tests.py (hunk 7 above)
Comment (MishNajam) [reply behave/features/steps/fala_end_to_end_tests.py:128]: the parenthesis has been added by the pre-commit linting hooks. The line is too long when placed on the same line.

File: behave/features/steps/fala_end_to_end_tests.py (hunk 6 above)
Comment (MishNajam) [reply behave/features/steps/fala_end_to_end_tests.py:133]: Oof, good spot, added.

**Review by psweeting1 — APPROVED** [review]
LGTM

File: behave/features/cla_fala/fala_end_to_end_tests.feature [hunk 10]
```diff
@@ -27,8 +27,8 @@ Feature: FALA end to end tests
     Then the result page containing "<location>" is updated to apply the filter "<filter_label>"
     Examples:
       | location | filter_label |
-      | SW1H9AJ  | Crime        |
-      | London   | Housing      |
+      | SW1H9AJ  | crm          |
```
Comment (jogilder) [inline behave/features/cla_fala/fala_end_to_end_tests.feature:30]: why are these changed?

Comment (MishNajam) [reply behave/features/cla_fala/fala_end_to_end_tests.feature:30]: The values of each of the checkbox items are linked to a 3/4 letter string for each of them instead of the readable label itself. 
Example: the label may read "Claims Against Pubic Authorities" but the checkbox input value assigned is "aap" 

FALA also uses this same 3/4 letter string to build the URL and ignores the label. For example the url generated from above step would add "&categories=aap" instead of the nice lovely checkbox. 

 I'm not sure whether this is django or if someone has intentionally designed it this way.

**Review by Pheonnexx — APPROVED** [review]

**Review by jogilder — CHANGES_REQUESTED** [review]

File: behave/features/cla_fala/fala_end_to_end_tests.feature (hunk 10 above)
Comment (jogilder) [reply behave/features/cla_fala/fala_end_to_end_tests.feature:30]: Discussed and understood the reason for this.

File: behave/features/steps/fala_end_to_end_tests.py [hunk 11]
```diff
@@ -68,18 +77,21 @@ def step_impl_apply_filter(context):
     'the result page containing "{location}" is updated to apply the filter "{filter_label}"'
 )
 def step_impl_update_result_page(context, location, filter_label):
-    current_path = context.helperfunc.get_current_path()
-    title_xpath = context.helperfunc.find_by_xpath("//html/body/div/main/div/div/h1")
+    current_url = context.helperfunc.driver().current_url
+    title_xpath = context.helperfunc.find_by_xpath(
```
[the remark this answers is not in the document]
Comment (jogilder) [reply behave/features/steps/fala_end_to_end_tests.py:81]: @MishNajam agreed to take this on board and refactor in the next ticket which is also on fala end to end

File: behave/features/steps/fala_end_to_end_tests.py [hunk 12]
```diff
@@ -88,4 +100,38 @@ def step_impl_update_result_page(context, location, filter_label):
...
+
+@step(
+    'I am taken to the page corresponding to the "{location}" "{organisation}" search result'
+)
+def step_impl_result_page_with_multi_params(context, location, organisation):
+    result_number_paragraph = context.helperfunc.find_by_xpath(
+        '//p[@class="govuk-body"]'
+    )
+    current_url = context.helperfunc.driver().current_url
+    title_xpath = context.helperfunc.find_by_xpath(
```
[the remark this answers is not in the document]
Comment (jogilder) [reply behave/features/steps/fala_end_to_end_tests.py:115]: @MishNajam agreed to take this on board and refactor in the next ticket which is also on fala end to end

**Review by jogilder — APPROVED** [review]
@MishNajam agreed to take this on board and refactor in the next ticket which is also on fala end to end
